# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнила студентка гр. 0392 Сидорина Дарья, вариант 85.


## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов​.

## !Основные теоретические положения

Аддитивной цепочкой для $ n \in ℕ $ называется последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-то двух предыдущих:
$$ a_i = a_j + a_k, \quad k \le j \le i, \quad i = 1, 2, ..., r $$

$l(n) = r$ - наименьшая длина аддитивной цепочки для $ n \in ℕ $.

Для метода наименьших множителей: $ \quad l(mn) \le l(m) + l(n) $

Для m-арного метода: $ \quad l(n) \le m - 2 + (k + 1)t \quad [m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j}] $

Для SX-метода: $ \quad l(n) \le \lambda (n) + \nu (n) - 1 $


### Свойства аддитивных цепочек:
* Полагается строгое возрастание элементов цепочки:
$ 1 = a_0 < a_1 < a_2 < ... < a_n = n $
* Одинаковые числа в цепочке можно опустить
* Пара $ (j, k), 0 \le k \le j < i $ называется шагом $i$
* Если $\exists$ более чем 1 пара $ (j, k) $, полагаем $ max \hspace{0.2cm} j $

### Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $ (линейный шаг)
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $, где $ \lambda (n) = \lfloor lb(n) \rfloor$

### Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым

### Теорема:

Если аддитивная цепочка содержит $d$ и $f = r - d$ неудвоений, то $n \le 2^{d-1} F_{f+3}$, где $F_j$ - число Фиббоначи

### Следствие:

Если аддитивная цепочка содержит $f$ удвоений и $S$ малых шагов, то

${S \le f \le} {S \over {1 - lb(\varphi)}}$, где $\varphi = {{\sqrt{5} + 1} \over 2} $ - золотое сечение

### Алгоритм Брауэра:

Для $n \in ℕ$ при заданном $k \in ℕ$ можно построить цепочку Брауэра с помощью рекуррентной формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Данная цепочка имеет длину
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимализирована для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

**Ход алгоритма:**

* Задаётся некий параметр $k$ для $n$.
Выполняется вычисление вспомогательных чисел:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d => n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d => q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d,$ следовательно $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ..., 2^k - 1, \\
2q_s, 4q_s, 8q_s, ..., 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ..., 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
..., \\
..., 2^k q_1, 2^k q_1 + r_1 = n.$$


## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.


## Выполнение работы

**1) Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для $𝑥^𝑛$ для 2-3 значений 𝑛 (значение 𝑛 > 30 выбираются студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.**


a) $n=33$

* Бинарный метод: 
$n = 33 = 32 + 1 = 100001_2$
<pre>
| (x^1)^2 * x^0 = x^2   | ---> 1 операция |
| (x^2)^2 * x^0 = x^4   | ---> 1 операция |
| (x^4)^2 * x^0 = x^8   | ---> 1 операция |
| (x^8)^2 * x^0 = x^16  | ---> 1 операция |
| (x^16)^2 * x^1 = x^33 | ---> 2 операции |
|-----------------------------------------|
|Всего: 6 операций|
</pre>


* Метод множителей: 
$n = 33 = 3*11$
следовательно, x^33 = (x^3)^11

<pre>
| x, x^2, x^3(=y)                | ---> 2 операции |
| y, y^2, y^4, y^5, y^10, y^11   | ---> 5 операций |
|--------------------------------------------------|
|Всего: 2+5=7 операций|
</pre>
<br>

Вывод: бинарный метод возводит в степень за меньшее число операций, чем метод множителей.

<br>
<br>
<br>


б) $n=64$

* Бинарный метод: 
$n = 64 = 1000000_2$
<pre>
| (x^1)^2 * x^0 = x^2   | ---> 1 операция |
| (x^2)^2 * x^0 = x^4   | ---> 1 операция |
| (x^4)^2 * x^0 = x^8   | ---> 1 операция |
| (x^8)^2 * x^0 = x^16  | ---> 1 операция |
| (x^16)^2 * x^0 = x^32 | ---> 1 операция |
| (x^32)^2 * x^0 = x^64 | ---> 1 операция |
|-----------------------------------------|
|Всего: 6 операций|
</pre>


* Метод множителей: 
$n = 64 = 2*32$
следовательно, x^64 = (x^2)^32

<pre>
| x, x^2(=y)                     | ---> 1 операция |
| y, y^2, y^4, y^8, y^16, y^32   | ---> 5 операций |
|--------------------------------------------------|
|Всего: 1+5=6 операций|
</pre>
<br>

Вывод: оба метода возводят в степень за одинаковое число операций.

<br>
<br>
<br>
<br>
в) $n=63$

* Бинарный метод: 
$n = 63 = 111111_2$
<pre>
| (x^1)^2 * x^1 = x^3   | ---> 2 операции |
| (x^3)^2 * x^1 = x^7   | ---> 2 операции |
| (x^7)^2 * x^1 = x^15  | ---> 2 операции |
| (x^15)^2 * x^1 = x^31 | ---> 2 операции |
| (x^31)^2 * x^1 = x^63 | ---> 2 операции |
|-----------------------------------------|
|Всего: 10 операций|
</pre>


* Метод множителей: 
$n = 63 = 3*21$
следовательно, x^63 = (x^3)^21

<pre>
| x, x^2, x^3(=y)                      | ---> 2 операции |
| y, y^2, y^4, y^5, y^10, y^20, y^21   | ---> 6 операций |
|--------------------------------------------------------|
|Всего: 2+6=8 операций|
</pre>
<br>


Вывод: метод множителей возводит в степень за меньшее число операций, чем бинарный метод, так как 63 в своей двоичной записи имеет максимальное число единиц.

<br>


Анализ:
Получается, что для чисел с небольшим количеством единиц в двоичной записи бинарный метод работает эффективнее, чем метод множителей. Однако в среднем метод множителей показываем результат лучше, чем бинарный метод. Метод множителей показывает свой наилучший результат при $n = 2^k -1 $
<br>
<br>

**2) Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра 𝑘, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.**

In [40]:
# Алгоритм Брауэра
def Brauer(n,k):
    chain = [] #цепочка
    remainder = [] #остатки
    d = 2**k # вспомогательные числа
    q = n
    
    #раскладываем n по модулю 2^k
    remainder.append([n, 0])
    while(q//d >= 1):
        remainder.append([q//d, q%d])
        q//=d
    remainder = remainder[::-1]
    
    #записываем первые элементы цепочки от 1 до qs < d
    for i in range(1, remainder[0][0]+1):
        chain.append(i)
        
    p = 1
    while(chain[-1] != n):
        #Удваиваем последний элемент цепочки, если результат не будет превосходить qn
        if(chain[-1] + chain[-1] <= remainder[p][0]):
            chain.append(chain[-1] + chain[-1])
            #либо прибавляем к последнему элементу остаток по модулю rn
        else:
            chain.append(chain[-1] + remainder[p - 1][1])
            p += 1
    return chain

In [41]:
#на проверку
n = 975
k = 4
check = Brauer(n, k)
print(check)
print("Длина цепочки равна: ", len(check))

[1, 2, 3, 6, 12, 24, 48, 60, 120, 240, 480, 960, 975]
Длина цепочки равна:  13


    [1, 2, 3, 6, 12, 24, 48, 60, 120, 240, 480, 960, 975]
    Длина цепочки равна:  13

In [42]:
# проаналиируем зависимоать длины цепочки от k
# N менять
N = 975
check = []
for i in range(2, 30):
    check.append(len(Brauer(N, i)))
print("min len is", min(check), "assuming k =", check.index(min(check)) + 2)

n = 1618033992434567890309877765443276543456784345678903456789098765432567
kk = math.log2(math.log2(n)) - math.log2(math.log2(math.log2(n)))
check = []
for i in range(2, 10):
    check.append(len(Brauer(n, i)))
print("min len is", min(check), "assuming k =", check.index(min(check)) + 2)
print("theory says that the minimum is reached assuming k =", kk)

for k in range(2, 14):
    fst6 = []
    for i in range(5000, 50000):
        fst6.append(len(Brauer(i, k)))
    print("k =", k, "min len is", min(fst6), "average aver =", sum(fst6)/len(fst6))

min len is 11 assuming k = 9
min len is 269 assuming k = 6
theory says that the minimum is reached assuming k = 4.873151751064757
k = 2 min len is 18 average aver = 21.66591111111111
k = 3 min len is 16 average aver = 20.3182
k = 4 min len is 15 average aver = 21.151022222222224
k = 5 min len is 15 average aver = 25.719155555555556
k = 6 min len is 14 average aver = 20.19788888888889
k = 7 min len is 16 average aver = 36.12215555555556
k = 8 min len is 28 average aver = 115.91786666666667
k = 9 min len is 19 average aver = 63.209244444444444
k = 10 min len is 15 average aver = 37.354933333333335
k = 11 min len is 14 average aver = 24.927155555555554
k = 12 min len is 14 average aver = 19.213266666666666
k = 13 min len is 14 average aver = 483.6913333333333


    min len is 11 assuming k = 9
    min len is 269 assuming k = 6
    theory says that the minimum is reached assuming k = 4.873151751064757
    k = 2 min len is 18 average aver = 21.66591111111111
    k = 3 min len is 16 average aver = 20.3182
    k = 4 min len is 15 average aver = 21.151022222222224
    k = 5 min len is 15 average aver = 25.719155555555556
    k = 6 min len is 14 average aver = 20.19788888888889
    k = 7 min len is 16 average aver = 36.12215555555556
    k = 8 min len is 28 average aver = 115.91786666666667
    k = 9 min len is 19 average aver = 63.209244444444444
    k = 10 min len is 15 average aver = 37.354933333333335
    k = 11 min len is 14 average aver = 24.927155555555554
    k = 12 min len is 14 average aver = 19.213266666666666
    k = 13 min len is 14 average aver = 483.6913333333333

**Анализ:**
Видно, что оптимальное $k(n)$ совсем невелико по сравнению с $n$, более точно оно определяется как
$$ k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n) $$ для достаточно больших $n$.<br>
В данном случае $n$ были недостаточно большими: оптимальное $k$ не превышает 1, при этом на практике оно приближённо равно 2: видно, что и для первых пяти чисел в диапазоне $[17, 127]$, и для последних пяти чисел в диапазоне $[127, 65535]$ $l_B(n)$ только увеличивается при увеличении $k \ge 2$.<br>
Кроме того, слишком большое $k$ ведёт к деградированию цепочек, построенных алгоритмом Брауэра, до рядов натуральных чисел. Это можно проследить в первую очередь на меньших числах: при $k = 6$ это $n = 18,49$, при $k = 9$ это все выбранные числа $n = 18, 49, 92, 94, 103$.<br>
Очевидно, полное совпадение цепочки метода Брауэра с рядом натуральных чисел происходит при таком $k$, что $2^k \ge n$. Это можно увидеть при $k = 16$, ведь $2^{16} = 65536$, что больше любого выбранного $n$.<br>
Однако, увеличение длины цепочки для чисел из второго диапазона $[127, 65535]$ уже заметно и при $k = 9$: например, для $n = 50597$ $l_B(n) = 429$, в то время как минимальная $l(n) = 21$.<br>

Таким образом, чтобы алгоритм Брауэра строил аддитивные цепочки адекватной длины, $k$ должно быть куда меньше $n$ и для достаточно больших $n$ вычисляться по формуле $ k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n) $, а для небольших $n$ быть положено равным $1$ или $2$ (разница должна быть не очень заметной).В результате исследований замечено, что длина цепочки сильно зависит от выбранного k. Более того, метод оптимального выбора k для достаточно больших n: $k = λλ(n) - 2λλλ(n)$  не оправдал себя. Его результаты при вменяемых n меньше едниницы, а при действительно больших n не больше 2. как правило, минимум, полученный переборм, больше.
Среднее значение k, при коротом достигается минимум, колеблется от 5 (для больших чисел) до 15 (для относительно небольших.) Тем не менее, пренебрегая незначительными различиями в длинне внутри одного порядка, можно сказать, что оптимальное $k \in [3, 8]$

**3) Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений 𝑛 > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.**


**4) Проверить гипотезу Шольца–Брауэра для всех натуральных 1<= 𝑛 <=10 на алгоритме дробления вектора индексов. Сделать выводы.**

## Выводы

В ходе выполнения данной практической работы были сформированы представления об аддитивных цепочках, были выработаны умения реализовывать и применять алгоритмы для нахождения минимальных аддитивных цепочек, был сформирован навык использования системы компьютерной математики SageMath для реализации таких алгоритмов.

На практике была проверена гиптеза Шольца-Брауэра, были проанализированы алгоритм Брауэра, бинарный метод и метод множителей.